opt: Add Ordering physical property

andy-kimball opened the pull request:
The Ordering physical property requires result rows to be sorted
according to a desired key. This is usually the result of an ORDER BY
expression, but can also be a requirement imposed by an operator like
merge join. If an expression can provide the required ordering (e.g.
a Scan operator that returns rows in primary key order), then nothing
more is needed. If not, then the optimizer will insert a Sort enforcer
into the expression tree to provide the ordering.

The memo group now keeps a "best expression map", which remembers the
lowest cost expression for that group, that also provides each required
set of properties. As various alternatives are explored, the best expr
is updated as lower cost expressions are found. The optimizer now
traverses the whole expression tree in order to find best expression(s)
for each group and to insert enforcers where needed.

This PR also contains a bare-bones implementation of ORDER BY for
testing purposes; it will need to be fully implemented later.

Release note: None

andy-kimball committed:
opt: Add Ordering physical property

The Ordering physical property requires result rows to be sorted
according to a desired key. This is usually the result of an ORDER BY
expression, but can also be a requirement imposed by an operator like
merge join. If an expression can provide the required ordering (e.g.
a Scan operator that returns rows in primary key order), then nothing
more is needed. If not, then the optimizer will insert a Sort enforcer
into the expression tree to provide the ordering.

The memo group now keeps a "best expression map", which remembers the
lowest cost expression for that group, that also provides each required
set of properties. As various alternatives are explored, the best expr
is updated as lower cost expressions are found. The optimizer now
traverses the whole expression tree in order to find best expression(s)
for each group and to insert enforcers where needed.

This PR also contains a bare-bones implementation of ORDER BY for
testing purposes; it will need to be fully implemented later.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23123)
<!-- Reviewable:end -->

RaduBerinde commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/opt/physical_props.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6JmkeELGpS1vJVHvll:-L6JmkeELGpS1vJVHvlm:b9zzja6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a48b4e1de7be7cfea3c3c4bb3f6be1dc530939d/pkg/sql/opt/opt/physical_props.go#L197)):*
> ```Go
> 
> // Provides returns true iff this ordering is a prefix of the given ordering.
> func (o Ordering) Provides(required Ordering) bool {
> ```

What is your thinking on how we will handle logical properties like column equivalencies or constant columns? E.g. if we have an Ordering a+,b+ but there is a filter `a = 1`, the "reduced" ordering is just `b+`. It seems that we would need to pass logical props to functions like this one.. And then the question is, do we want to reduce the ordering each time, or should we always store it in reduced form?

---

*[pkg/sql/opt/xform/optimizer.go, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Jo05G4PA0jlzN2KDX:-L6Jo05G4PA0jlzN2KDY:b-j6e55v) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a48b4e1de7be7cfea3c3c4bb3f6be1dc530939d/pkg/sql/opt/xform/optimizer.go#L99)):*
> ```Go
> // the following SQL query:
> //
> //   SELECT * FROM a WHERE y=1 ORDER BY y
> ```

I think you meant `WHERE x=1`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/opt/opt/physical_props.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6JmkeELGpS1vJVHvll:-L6JyCZcsMAqYewrl3Jr:b-5b7tra) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a48b4e1de7be7cfea3c3c4bb3f6be1dc530939d/pkg/sql/opt/opt/physical_props.go#L197)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What is your thinking on how we will handle logical properties like column equivalencies or constant columns? E.g. if we have an Ordering a+,b+ but there is a filter `a = 1`, the "reduced" ordering is just `b+`. It seems that we would need to pass logical props to functions like this one.. And then the question is, do we want to reduce the ordering each time, or should we always store it in reduced form?
</blockquote></details>

I'll have to think about it more. My preliminary thoughts are that the case you bring up shows that the current `Ordering` prop may not be enough. Maybe we'd have a way to specify the columns that are part of the "equality prefix", like: `o:(+a)+b`. The parentheses signify that `a` does not need to be part of the provided ordering. Operators receiving this requirement could choose to provide `o:+a+b`, `o:+a` or `o:+b`. All would be acceptable to the parent.

---

*[pkg/sql/opt/xform/optimizer.go, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Jo05G4PA0jlzN2KDX:-L6Jx0jFIKGnPfEp1QvV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a48b4e1de7be7cfea3c3c4bb3f6be1dc530939d/pkg/sql/opt/xform/optimizer.go#L99)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think you meant `WHERE x=1`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/opt/physical_props.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6JmkeELGpS1vJVHvll:-L6MbRrfXdzt8WC6wgsv:b-b2wo4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a48b4e1de7be7cfea3c3c4bb3f6be1dc530939d/pkg/sql/opt/opt/physical_props.go#L197)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'll have to think about it more. My preliminary thoughts are that the case you bring up shows that the current `Ordering` prop may not be enough. Maybe we'd have a way to specify the columns that are part of the "equality prefix", like: `o:(+a)+b`. The parentheses signify that `a` does not need to be part of the provided ordering. Operators receiving this requirement could choose to provide `o:+a+b`, `o:+a` or `o:+b`. All would be acceptable to the parent.
</blockquote></details>

correction: `o:+a` would not be an acceptable ordering in this case, just the other two. After thinking about it more, I do think that this would be a good way to solve the problem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 15 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/optimizer.go, line 233 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MzidzBAClyW7uhmCh:-L6MzidzBAClyW7uhmCi:bo4n3ew) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/optimizer.go#L233)):*
> ```Go
> //              ├── variable: a.x [type=int]
> //              └── const: 1 [type=int]
> //
> ```

This was a nice read.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 13 of 14 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/opt/physical_props.go, line 196 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Ml6eB5KeNmyJdz0U0:-L6Ml6eB5KeNmyJdz0U1:blsbcap) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/opt/physical_props.go#L196)):*
> ```Go
> }
> 
> // Provides returns true iff this ordering is a prefix of the given ordering.
> ```

I think this is backwards - did you mean "iff the given required ordering is a prefix of this ordering"?

---

*[pkg/sql/opt/optbuilder/builder.go, line 542 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mm-Bub2sfC6k6SlHU:-L6Mm-Bub2sfC6k6SlHV:b9p69fs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/optbuilder/builder.go#L542)):*
> ```Go
> 	// correctly handle cases like `SELECT a FROM t ORDER BY c`. The ordered
> 	// property can only contain refs to output columns, so the `c` column
> 	// must be retained in the projection (and presentation property then omits
> ```

This is helpful - thanks!

---

*[pkg/sql/opt/optbuilder/builder.go, line 552 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MmIxsfoZ9bmx9zKLS:-L6MmIxsfoZ9bmx9zKLT:b-3zo954) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/optbuilder/builder.go#L552)):*
> ```Go
> 
> // buildSelectClause builds a set of memo groups that represent the given
> // select clause. We pass the entire select statement rather than just the
> ```

Seems like we don't need the whole statement anymore. I think you can just pass in the `SelectClause` (and update the comment).

---

*[pkg/sql/opt/optbuilder/builder.go, line 1339 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MmkcZVana0mJlr8Jq:-L6MmkcZVana0mJlr8Jr:b-xonv4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/optbuilder/builder.go#L1339)):*
> ```Go
> 	}
> 
> 	// Order-by is not a relational expression,
> ```

This comment doesn't really explain what comes next...

---

*[pkg/sql/opt/xform/memo.go, line 182 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mpm1QiNpyRvRhq2UV:-L6Mpm1QiNpyRvRhq2UW:b-ye0hn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo.go#L182)):*
> ```Go
> 		id:           id,
> 		exprs:        exprs,
> 		bestExprsMap: make(map[opt.PhysicalPropsID]int),
> ```

FastIntMap?

---

*[pkg/sql/opt/xform/memo_group.go, line 49 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MsIrUwivFk0eoVWh2:-L6MsIrUwivFk0eoVWh3:b-ye0hn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo_group.go#L49)):*
> ```Go
> 	// bestExprs remembers the lowest cost expression in the group that provides
> 	// a particular set of physical properties.
> 	bestExprsMap map[opt.PhysicalPropsID]int
> ```

FastIntMap?

---

*[pkg/sql/opt/xform/memo_group.go, line 70 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MsfVXdPG1g2MYwKwZ:-L6MsfVYoffLoxZX7JBI:b86m253) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo_group.go#L70)):*
> ```Go
> // ensureBestExpr looks up the bestExpr that has the lowest cost for the given
> // required properties. If no bestExpr exists yet, then ensureBestExpr creates
> // an empty bestExpr and returns it.
> ```

Change to: "... creates an empty bestExpr, adds it to bestExprsMap, and returns it."

---

*[pkg/sql/opt/xform/optimizer.go, line 389 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mx-JSBNGBYCntmTrr:-L6Mx-JSBNGBYCntmTrs:b4ymdz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/optimizer.go#L389)):*
> ```Go
> 	// Check whether this is the new lowest cost expression with the enforcer
> 	// added. Note that the enforcer is not in the group's exprs slice, so its
> 	// memoLoc.expr field is 0.
> ```

Will this be a problem if there are multiple enforcers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-:-L6N-NPW7Ple_szsrESi:b-uwkowd)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM, good stuff! I will make another pass, especially over the tests to make sure I didn't miss any details.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/opt/physical_props.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6JmkeELGpS1vJVHvll:-L6MxTK8ixSX_kKBtTwG:b-akmoq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a48b4e1de7be7cfea3c3c4bb3f6be1dc530939d/pkg/sql/opt/opt/physical_props.go#L197)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

correction: `o:+a` would not be an acceptable ordering in this case, just the other two. After thinking about it more, I do think that this would be a good way to solve the problem.
</blockquote></details>

There are other ways in which the logical props can reduce orderings. (Especially if we add general FDs at some point). For example, if we know that a = c, then `+a,+b,+c,+d` is the same with `+a,+b`. Or if we know that `a,c` is a key, it is the same with `+a,+b,+c`.

I would think that instead of putting more of this information into Ordering, we would consult the logical props in conjunction with it.

---

*[pkg/sql/opt/optbuilder/builder.go, line 552 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mz41wzDcvAAW8wFQd:-L6Mz41wzDcvAAW8wFQe:b-fxsktk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/optbuilder/builder.go#L552)):*
> ```Go
> 
> // buildSelectClause builds a set of memo groups that represent the given
> // select clause. We pass the entire select statement rather than just the
> ```

This comment needs updating. Perhaps we can now pass the `SelectClause`?

---

*[pkg/sql/opt/xform/best_expr.go, line 22 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MzOt2DlbFfIT1G2aH:-L6MzOt2DlbFfIT1G2aI:b1wkdzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/best_expr.go#L22)):*
> ```Go
> )
> 
> type bestExpr struct {
> ```

[nit] A comment on the type would help. Maybe bestExprContainer?

---

*[pkg/sql/opt/xform/best_expr.go, line 32 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mza9s1YJXIrqtEJrx:-L6Mza9s1YJXIrqtEJry:b-1y4rbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/best_expr.go#L32)):*
> ```Go
> 	fullyOptimized bool
> 
> 	// fullyOptimizedExprs contains the set of expressions that have been fully
> ```

[nit] mention that it's a set of exprIDs

---

*[pkg/sql/opt/xform/memo.go, line 365 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6N-2Mye754GRqJ_5TG:-L6N-2Mye754GRqJ_5TH:bi3g6j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo.go#L365)):*
> ```Go
> 		buf.Reset()
> 
> 		// Don't show best expression for scalar groups because they're not too
> ```

[nit] expressions

---

*[pkg/sql/opt/xform/memo_group.go, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6N-aV4gSvAxG6oSDTq:-L6N-aV4gSvAxG6oSDTr:bf9248x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo_group.go#L60)):*
> ```Go
> // lookupBestExpr looks up the bestExpr that has the lowest cost for the given
> // required properties. If no bestExpr exists yet, lookupBestExpr returns nil.
> func (g *memoGroup) lookupBestExpr(required opt.PhysicalPropsID) *bestExpr {
> ```

[nit] do we need to provide some guidance for how to use the `*bestExpr` if we're modifying it? (specifically that we shouldn't modify it after a call to `ensureBestExpr`)

---

*[pkg/sql/opt/xform/optimizer.go, line 233 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MzidzBAClyW7uhmCh:-L6N0qAv5a5X6sENEO3U:be8ohqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/optimizer.go#L233)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This was a nice read.
</blockquote></details>

Agreed, very nice description, thanks!

A case I'm wondering about is this: say we have `ORDER BY a,b` and we have two covering indexes, one on `a` and another one on something else. The latter is cheaper to scan (e.g. it has fewer columns). The best plan involves using the index on `a` and sorting it (it's much cheaper to sort, as we only sort groups of rows for each value of `a`). The way the algorithm is laid out, the enforcer would require nothing of its input and would always choose the other index. It seems that we would need to try various "trimmed down" versions of the required properties (in this case we would try to require nothing, or order by `a`). (which doesn't sound too bad).

But this is kind of a more general thing with the top-down approach. A more difficult case is planning merge-joins: we have two inputs that we might be able to cheaply get sorted in a way that we can do mergejoin; but from top-down there are many possible permutations of orderings and it doesn't make sense to try them all (it would make more sense to somehow get an idea of what kind of orderings you could get from below). For example: `abcd` NATURAL JOIN `abce`. We could do merge-join if both sides have the same ordering on `a,b,c` but there are 48 such orderings!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-:-L6N4A_KaWb3vmyipgir:b-lc0928)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 10 of 15 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/opt/physical_props.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6JmkeELGpS1vJVHvll:-L6NZ1oBI7uj89DodTGj:b-e75n4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a48b4e1de7be7cfea3c3c4bb3f6be1dc530939d/pkg/sql/opt/opt/physical_props.go#L197)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

There are other ways in which the logical props can reduce orderings. (Especially if we add general FDs at some point). For example, if we know that a = c, then `+a,+b,+c,+d` is the same with `+a,+b`. Or if we know that `a,c` is a key, it is the same with `+a,+b,+c`.

I would think that instead of putting more of this information into Ordering, we would consult the logical props in conjunction with it.
</blockquote></details>

Radu and I talked about this offline. We both agree that there will be additional information needed as we increase the sophistication of the optimizer, whether it be a more complex `Ordering` and/or consulting logical props when deriving required props.

---

*[pkg/sql/opt/opt/physical_props.go, line 196 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Ml6eB5KeNmyJdz0U0:-L6NLGR76HDS5D6v4TEQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/opt/physical_props.go#L196)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think this is backwards - did you mean "iff the given required ordering is a prefix of this ordering"?
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/builder.go, line 552 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MmIxsfoZ9bmx9zKLS:-L6NZPcMfAeJN885mADP:buzo7gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/optbuilder/builder.go#L552)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Seems like we don't need the whole statement anymore. I think you can just pass in the `SelectClause` (and update the comment).
</blockquote></details>

Actually, I think we still need that. I deleted the `OrderBy` handling below b/c lint wouldn't let me keep it once I removed the `panic`. But it will have to be added back. I'll put comment back in, at least.

---

*[pkg/sql/opt/optbuilder/builder.go, line 552 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mz41wzDcvAAW8wFQd:-L6NbqN8BcHmsbmcKA9K:b-6gwbgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/optbuilder/builder.go#L552)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This comment needs updating. Perhaps we can now pass the `SelectClause`?
</blockquote></details>

See my answer to Rebecca above - I added back TODO comment.

---

*[pkg/sql/opt/optbuilder/builder.go, line 1339 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MmkcZVana0mJlr8Jq:-L6NbwQbuZ2tv4qheefS:b-cfx4uu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/optbuilder/builder.go#L1339)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This comment doesn't really explain what comes next...
</blockquote></details>

This was to keep you in suspense. : )
Removed that, since it's part of func comment.

---

*[pkg/sql/opt/xform/best_expr.go, line 22 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MzOt2DlbFfIT1G2aH:-L6NdDrHfk6r3IsAxo1F:b-ughw63) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/best_expr.go#L22)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] A comment on the type would help. Maybe bestExprContainer?
</blockquote></details>

I added a comment. Trouble with `bestExprContainer` is that it's loooong, and I don't think the extra length clarifies it enough to be worth it. Hopefully the comment (and other comments) clarify it enough that it's not a problem.

---

*[pkg/sql/opt/xform/best_expr.go, line 32 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mza9s1YJXIrqtEJrx:-L6Ne1IhlbJ27FBvj5mu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/best_expr.go#L32)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] mention that it's a set of exprIDs
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/memo.go, line 182 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mpm1QiNpyRvRhq2UV:-L6NgRJbblOmG4Kx36nU:b-q29a2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo.go#L182)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

FastIntMap?
</blockquote></details>

Good suggestion, I switched it to use FastIntMap.

---

*[pkg/sql/opt/xform/memo.go, line 365 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6N-2Mye754GRqJ_5TG:-L6NgZIanqlIq5POkr_V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo.go#L365)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] expressions
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/memo_group.go, line 49 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MsIrUwivFk0eoVWh2:-L6NgdJZH2xyc8kmQqmq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo_group.go#L49)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

FastIntMap?
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/memo_group.go, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6N-aV4gSvAxG6oSDTq:-L6NnUZy3wDist-ccBVv:b-ppsj0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo_group.go#L60)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] do we need to provide some guidance for how to use the `*bestExpr` if we're modifying it? (specifically that we shouldn't modify it after a call to `ensureBestExpr`)
</blockquote></details>

But we do modify it after calling `ensureBestExpr` by calling `ratchetCost` on it. Can you suggest some language for what you mean?

---

*[pkg/sql/opt/xform/memo_group.go, line 70 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MsfVXdPG1g2MYwKwZ:-L6NhBCmPqIZJHrE1rvY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo_group.go#L70)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Change to: "... creates an empty bestExpr, adds it to bestExprsMap, and returns it."
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/optimizer.go, line 233 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MzidzBAClyW7uhmCh:-L6NjRsWghIkB5TaLO2l:bu6t1vc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/optimizer.go#L233)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Agreed, very nice description, thanks!

A case I'm wondering about is this: say we have `ORDER BY a,b` and we have two covering indexes, one on `a` and another one on something else. The latter is cheaper to scan (e.g. it has fewer columns). The best plan involves using the index on `a` and sorting it (it's much cheaper to sort, as we only sort groups of rows for each value of `a`). The way the algorithm is laid out, the enforcer would require nothing of its input and would always choose the other index. It seems that we would need to try various "trimmed down" versions of the required properties (in this case we would try to require nothing, or order by `a`). (which doesn't sound too bad).

But this is kind of a more general thing with the top-down approach. A more difficult case is planning merge-joins: we have two inputs that we might be able to cheaply get sorted in a way that we can do mergejoin; but from top-down there are many possible permutations of orderings and it doesn't make sense to try them all (it would make more sense to somehow get an idea of what kind of orderings you could get from below). For example: `abcd` NATURAL JOIN `abce`. We could do merge-join if both sides have the same ordering on `a,b,c` but there are 48 such orderings!
</blockquote></details>

Yes, I agree that we could add more logic to deciding which physical property combos to require, by using bottom-up props to decide on what top-down props to require. Definitely something to continue to think about.

---

*[pkg/sql/opt/xform/optimizer.go, line 389 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Mx-JSBNGBYCntmTrr:-L6NjpqhGWEMdXok6AiV:b-36c4b6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/optimizer.go#L389)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Will this be a problem if there are multiple enforcers?
</blockquote></details>

Should work fine. Opttoy had multiple enforcers, back when we had the `Present` enforcer. Each enforcer references its own group, but with a reduced set of phys props, so it doesn't self-reference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/optbuilder/builder.go, line 1339 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6MmkcZVana0mJlr8Jq:-L6NqHjzLKnNJmUfcLdj:bc52tdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/optbuilder/builder.go#L1339)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This was to keep you in suspense. : )
Removed that, since it's part of func comment.
</blockquote></details>

:laughing: 

---

*[pkg/sql/opt/xform/memo_group.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6NrAkFaz182Tk07_0e:-L6NrAkFaz182Tk07_0f:b3823cn) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e1de8a72d642695f69ea785323f3a3884b3d4f/pkg/sql/opt/xform/memo_group.go#L86)):*
> ```Go
> 
> // forEachBestExpr iterates over the bestExpr map and calls the given function
> // for each pair of
> ```

Comment got cut off

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/xform/memo.go, line 348 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Np_VBRbTyECWUrARA:-L6Np_VBRbTyECWUrARB:b-po7jxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e1de8a72d642695f69ea785323f3a3884b3d4f/pkg/sql/opt/xform/memo.go#L348)):*
> ```Go
> 	// Sort the bestExprs by required properties.
> 	beSort := make([]bestExprSort, len(mgrp.bestExprs))
> 	i := 0
> ```

[nit] Could append to `beSort` instead of maintaining `i`

---

*[pkg/sql/opt/xform/memo_group.go, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6N-aV4gSvAxG6oSDTq:-L6NoM6acJuwLbfZAuBW:b-cafeo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo_group.go#L60)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

But we do modify it after calling `ensureBestExpr` by calling `ratchetCost` on it. Can you suggest some language for what you mean?
</blockquote></details>

I mean - a call to `ensureBestExpr` potentially invalidates (at least for modification purposes) any `*bestExpr` reference previously obtained. You shouldn't obtain a `*bestExpr` (with either function), then cause the slice to potentially be reallocated (via an unrelated call to `ensureBestExpr`) and then modify the `*bestExpr` from step 1.

Maybe "the returned reference can only be used until the next call to ensureBestExpr".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 13 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt/xform/memo.go, line 348 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Np_VBRbTyECWUrARA:-L6OF49_4KEJpwnJUNJV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e1de8a72d642695f69ea785323f3a3884b3d4f/pkg/sql/opt/xform/memo.go#L348)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Could append to `beSort` instead of maintaining `i`
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/memo_group.go, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6N-aV4gSvAxG6oSDTq:-L6OEKKVnOaAKuXwig62:bsr48x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad4cf0965f117c88ab3961e245049eac58bbe2dd/pkg/sql/opt/xform/memo_group.go#L60)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I mean - a call to `ensureBestExpr` potentially invalidates (at least for modification purposes) any `*bestExpr` reference previously obtained. You shouldn't obtain a `*bestExpr` (with either function), then cause the slice to potentially be reallocated (via an unrelated call to `ensureBestExpr`) and then modify the `*bestExpr` from step 1.

Maybe "the returned reference can only be used until the next call to ensureBestExpr".
</blockquote></details>

Ah got it. I actually got caught by that a while back. I think it's worth that risk in order to allocate fewer objects, since this is in the inner loop of our optimizer. Definitely worth a warning to callers, though. Added.

---

*[pkg/sql/opt/xform/memo_group.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6NrAkFaz182Tk07_0e:-L6OEHLOnAffjBEePWXt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e1de8a72d642695f69ea785323f3a3884b3d4f/pkg/sql/opt/xform/memo_group.go#L86)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Comment got cut off
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/xform/memo_group.go, line 91 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6OGFaBoGoSr7-lNvOb:-L6OGFaBoGoSr7-lNvOc:bp735tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa9470ca19432c5351e84f6ab936d920ae18a53d/pkg/sql/opt/xform/memo_group.go#L91)):*
> ```Go
> // forEachBestExpr iterates over the bestExpr map and calls the given function
> // for each pair of (required property set, best expression).
> // NOTE: The returned bestExpr reference is only valid until the next call to
> ```

You're not returning a `bestExpr` here....

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/xform/memo_group.go, line 91 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6OGFaBoGoSr7-lNvOb:-L6OIRKNbp527hteawJQ:b-2xpho5) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa9470ca19432c5351e84f6ab936d920ae18a53d/pkg/sql/opt/xform/memo_group.go#L91)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You're not returning a `bestExpr` here....
</blockquote></details>

True, I'll rephrase to say the `bestExpr references passed to the function`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/optbuilder/select.go, line 152 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Oc8JyKjcS6dqeQHii:-L6Oc8JyKjcS6dqeQHij:b-gcm3j5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fd26402272dcaa902dd2808117d983421810c52/pkg/sql/opt/optbuilder/select.go#L152)):*
> ```Go
> 
> 	case *tree.ValuesClause:
> 		return b.buildValuesClause(t, inScope)
> ```

assign to `out, outScope`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/optbuilder/select.go, line 152 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123#-L6Oc8JyKjcS6dqeQHii:-L6OdMpa6srByG5PD3zm:blb7nf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fd26402272dcaa902dd2808117d983421810c52/pkg/sql/opt/optbuilder/select.go#L152)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

assign to `out, outScope`
</blockquote></details>

whoops, overlooked that during merge

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23123)*
<!-- Sent from Reviewable.io -->
